pyomo/common/dependencies.py

```diff
@@ -325,20 +392,31 @@ def _finalize_yaml(module, available):
 def _finalize_scipy(module, available):
     if available:
         # Import key subpackages that we will want to assume are present
+        import scipy.stats
         import scipy.sparse
         import scipy.spatial
-        import scipy.stats
 
 def _finalize_pympler(module, available):
     if available:
         # Import key subpackages that we will want to assume are present
         import pympler.muppy
 
+def _finalize_matplotlib(module, available):
+    if available:
+        import matplotlib.pyplot
+
 yaml, yaml_available = attempt_import('yaml', callback=_finalize_yaml)
 pympler, pympler_available = attempt_import(
     'pympler', callback=_finalize_pympler)
 numpy, numpy_available = attempt_import('numpy', alt_names=['np'])
-scipy, scipy_available = attempt_import('scipy', callback=_finalize_scipy)
+scipy, scipy_available = attempt_import(
+    'scipy', callback=_finalize_scipy, deferred_submodules={'stats':None})
 networkx, networkx_available = attempt_import('networkx', alt_names=['nx'])
 pandas, pandas_available = attempt_import('pandas', alt_names=['pd'])
 dill, dill_available = attempt_import('dill')
+
+# Note that matplotlib.pyplot can generate a runtime error on OSX when
+# not installed as a Framework (as is the case in the CI systems)
+matplotlib, matplotlib_available = attempt_import(
+    'matplotlib', only_catch_importerror=False, callback=_finalize_matplotlib,
+    deferred_submodules={'pyplot': ['plt']})
```

Comment:
If a developer doesn't follow the standard naming convention for this submodule (i.e. `import matplotlib.pyplot as pt`) what will happen? Will it just bypass the deferred import stuff altogether and eventually fail the test checking the import time?

Reply:
If the developer runs
```python
import matplotlib.pyplot as pt  # or even as plt
```
then this is a direct Python import: matplotlib and pyplot will be immediately imported.  If a user tries to do this from dependencies with something like:
```python
import pyomo.common.dependencies.matplotlib.pyplot as plt
```
they will get an import error.  Similarly, 
```python
from pyomo.common.dependencies import matplotlib.pyplot as plt
```
will get a parse error.  Finally (what I think you were originally asking), if they:
```python
from pyomo.common.dependencies import matplotlib
pt = matplotlib.pyplot
```
Then `pyplot` will not be imported until the user accesses an attribute on `pt`, at which point the module is imported and the attribute is returned.  The only difference between `plt` and `pt` is that because `pt` is not one of the known alternate names, then the deferred import occurs, the deferred importer cannot replace `pt` in the module `globals()` dictionary.  Subsequent calls to `pt.<any attribute>` will work just find, but will incur a slight overhead as they have to go through the `__getattr__` on `pt` (because `pt` is still an instance of `DeferredImportModule`)